### PR TITLE
[mdatagen] allow filtering out metrics based on resource attributes

### PR DESCRIPTION
**Pull request by povilasv**
**Description:** <Describe what has changed.>

This PR allows filtering out metrics based on resource attributes. For example:

```
        resource_attributes:                                                                      
          k8s.pod.name:
            enabled: true 
            exclude:
              #- strict: "kube-apiserver-kind-control-plane"
              - regexp: "kube-.*"
              - regexp: "coredns-.*"
              - strict: "coredns"
              - strict: "kindnet-mpb2p"
```

Would remove metrics that match regex or strict rules on resource attributes.

**Link to tracking Issue:** <Issue number if applicable>

https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/25134
**Testing:** <Describe what testing was performed and which tests were added.>
- Tested with k8scluster receiver in kind cluster.
- unit tests added
**Documentation:** <Describe the documentation added.>

### CODE REVIEW

File: cmd/mdatagen/validate.go [hunk 1]
```diff
@@ -109,6 +109,20 @@ func (md *metadata) validateResourceAttributes() error {
...
+				errs = multierr.Append(errs, fmt.Errorf(`resource attribute filter "%v": %w`, filter, err))
+				continue
+			}
+		}
+		for _, filter := range attr.Exclude {
+			if err := filter.Validate(); err != nil {
+				errs = multierr.Append(errs, fmt.Errorf(`resource attribute filter "%v": %w`, filter, err))
+				continue
+			}
+		}
```
Comment (dmitryax) [inline cmd/mdatagen/validate.go:124]: This is probably not needed. https://github.com/open-telemetry/opentelemetry-collector/blob/e7e73ab2755d9628089aa0cdaba8e2e68e31d604/component/config.go#L90 should take care of this. But please verify

File: internal/filter/filter/config.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+// CreateFilter creates a Filter from yaml config.
+func CreateFilter(configs []Config) (Filter, error) {
+	cf := &CombinedFilter{
+		filters: make(map[string]struct{}),
+	}
+	for _, config := range configs {
+		if err := config.Validate(); err != nil {
+			return nil, err
+		}
```
Comment (dmitryax) [inline internal/filter/filter/config.go:41]: Same here. Likely not needed

**Comment by dmitryax** [comment]
Good start. Thanks @povilasv!

File: internal/filter/filter/testdata/config.yaml [hunk 3]
```diff
@@ -0,0 +1,16 @@
+# Yaml form of the configuration for FilterSets
+# This configuration can be embedded into other component's yamls
+# The top level here are just test names and do not represent part of the actual configuration.
+
+regexp/default:
+    match_type: regexp
```
Comment (dmitryax) [inline internal/filter/filter/testdata/config.yaml:6]: Why `match_type`? I thought we implement the Option B from https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/25161

File: internal/filter/filter/filter.go [hunk 4]
```diff
@@ -0,0 +1,11 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package filter // import "github.com/open-telemetry/opentelemetry-collector-contrib/internal/filter/filter"
+
+// Filter is an interface for matching values against a set of filters.
+type Filter interface {
```
Comment (TylerHelmuth) [inline internal/filter/filter/filter.go:7]: `internal/filter` is already filled with interfaces (many which I'd like to remove). Do we definitely need to add another? Nothing in `expr` can be reused?

File: .chloggen/mdatagen-filter.yaml [hunk 5]
```diff
@@ -0,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: enhancement
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: "metadatagen"
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: "allow filtering attribute values using string and regexes"
```
Comment (fatsheep9146) [inline .chloggen/mdatagen-filter.yaml:10]: I suggest we'd better enhance this description, for example `support excluding some metrics based on string and regexes in resource_attributes`.

File: .chloggen/mdatagen-filter.yaml [hunk 6]
```diff
@@ -0,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: enhancement
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: "metadatagen"
```
Comment (fatsheep9146) [inline .chloggen/mdatagen-filter.yaml:7]: cmd/mdatagen

**Comment by fatsheep9146** [comment]
I think documentation.md should also be updated

File: internal/filter/filter/filter.go (hunk 4 above)
Comment (TylerHelmuth) [reply internal/filter/filter/filter.go:7]: I guess it is ok bc it is internal and we can easily update later.  Can we name it something more related to mdatagen/scrappers/config? `internal/filter/filter` isn't very descriptive.

Comment (TylerHelmuth) [reply internal/filter/filter/filter.go:7]: Also, filterset converts other types to string, can we do that?  What requires the use of `any` if we are doing regex/strict string matching?

Comment (TylerHelmuth) [reply internal/filter/filter/filter.go:7]: > I think the general idea is to have this filtering interface as the standard interface on receivers (not just mdatagen). It basically implements the Option B of this issue https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/25161

You're right, I forgot this was part of that issue.

File: internal/filter/filter/config.go [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+	cf := &CombinedFilter{
+		filters: make(map[any]struct{}),
+	}
+	for _, config := range configs {
+		/*
+			TODO: remove this once I check that Validate() is called automatically.
+				if err := config.Validate(); err != nil {
+					return nil, err
+				}
+		*/
```
Comment (dmitryax) [inline internal/filter/filter/config.go:51]: Can be removed :)

File: internal/filter/filter/config.go [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+type CombinedFilter struct {
+	filters map[any]struct{}
```
Comment (dmitryax) [inline internal/filter/filter/config.go:36]: Maybe `stricts`?

File: receiver/apachesparkreceiver/go.mod [hunk 9]
```diff
@@ -121,3 +122,7 @@ replace github.com/open-telemetry/opentelemetry-collector-contrib/pkg/pdatautil
 replace github.com/docker/docker v24.0.4+incompatible => github.com/docker/docker v24.0.5-0.20230719162248-f022632503d1+incompatible
 
 replace github.com/open-telemetry/opentelemetry-collector-contrib/pkg/golden => ../../pkg/golden
+
+replace github.com/open-telemetry/opentelemetry-collector-contrib/internal/filter => ../../internal/filter
+
+replace github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl => ../../pkg/ottl
```
Comment (dmitryax) [inline receiver/apachesparkreceiver/go.mod:128]: Why ottl?

File: receiver/couchdbreceiver/go.mod [hunk 10]
```diff
@@ -86,3 +87,9 @@ retract (
 )
 
 replace github.com/open-telemetry/opentelemetry-collector-contrib/pkg/golden => ../../pkg/golden
+
+replace github.com/open-telemetry/opentelemetry-collector-contrib/internal/filter => ../../internal/filter
+
+replace github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl => ../../pkg/ottl
+
+replace github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal => ../../internal/coreinternal
```
Comment (dmitryax) [inline receiver/couchdbreceiver/go.mod:95]: Is this added by `make crosslink`?

**Comment by dmitryax** [comment]
@povilasv PTAL at the failing CI

**Comment by dmitryax** [comment]
@povilasv the PR looks good to me in general. Can you please submit if against mdatagen in core?

**Pull request closed without merging**